arch/arm64/core/mmu.c

```diff
@@ -806,6 +806,79 @@ static uint64_t get_tcr(int el)
 	return tcr;
 }
 
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
+
+static int flush_inv_mapped_pages(int (*cmo_fn)(void *, size_t),
+				  uint64_t *pte, int level, uint64_t base)
+{
+	int ret = 0;
+	uint64_t level_size = BIT(LEVEL_TO_VA_SIZE_SHIFT(level));
+	const uint64_t max_va_addr = (uint64_t)CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE;
+
+	if (!xlat_use_count[table_index(pte)]) {
+		return 0;
+	}
+
+	for (unsigned int i = 0; i < Ln_XLAT_NUM_ENTRIES; i++) {
```

Comment:
This is incorrect for the topmost level, because topmost level can have less entries.

Or even more entries in case of concatenated tables.

Reply:
> could you please explain how it can happen with the current implementation of MMU mappings?

Of course: paragraph FTBXR of ARM DDI 0487I.a:  

> If TCR_ELx.TnSZ configures an IA size that is smaller than the maximum size, then each one-bit reduction in the
IA size has one of the following effects on the lookup level that the table walk starts with:
• The translation table size is reduced by one half. <==== This case
• The table walk starts one level later.

Again, this is valid only for the top-level table.

Reply:
Walking through code in this file I have seen only implementation that I have used moreover u-boot patch have exact the same size for loop but it is passed your review

Reply:
got it!

Reply:
> IA size has one of the following effects on the lookup level that the table walk starts with:
> • The translation table size is reduced by one half. <==== This case
> • The table walk starts one level later.

`one of the following effects `:
and for our case it is the second effect because we begin walk on level 1 (BASE_XLAT_LEVEL == 1)

pls, correct me if I am wrong